repository.go

```diff
@@ -805,6 +809,202 @@ func (r *Repository) resolveToCommitHash(h plumbing.Hash) (plumbing.Hash, error)
 	}
 }
 
+// Worktrees returns all the linked working trees in a repository.
+func (r *Repository) Worktrees() ([]*Worktree, error) {
+	type fsBased interface {
+		Filesystem() billy.Filesystem
+	}
+
+	fs, isFSBased := r.Storer.(fsBased)
+	if !isFSBased {
+		return nil, ErrRepositoryNotFileBased
+	}
+
+	common := fs.Filesystem()
+	files, err := common.ReadDir("worktrees")
+	if err != nil {
+		return nil, err
+	}
+
+	worktrees := []*Worktree{}
+	for _, file := range files {
+		f, err := common.Open(filepath.Join("worktrees", file.Name(), "gitdir"))
+		if os.IsNotExist(err) {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+
+		b, err := stdioutil.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+
+		path := strings.TrimSpace(string(b))
+		worktree, err := r.OpenWorktree(filepath.Dir(path))
+		if err != nil {
+			return nil, err
+		}
+		worktrees = append(worktrees, worktree)
+	}
+
+	return worktrees, nil
+}
+
+// CreateWorktree creates a linked working tree at the given path.
+// Specify the branch for working tree files with CheckoutOptions.
+func (r *Repository) CreateWorktree(path string, opts *CheckoutOptions) (*Worktree, error) {
+	type fsBased interface {
+		Filesystem() billy.Filesystem
+	}
+
+	fs, isFSBased := r.Storer.(fsBased)
+	if !isFSBased {
+		return nil, ErrRepositoryNotFileBased
+	}
+
+	_, err := os.Stat(filepath.Join(path, GitDirName))
+	if err == nil {
+		return nil, ErrWorktreeAlreadyExists
+	} else if !os.IsNotExist(err) {
+		return nil, err
+	}
+
+	name := filepath.Base(path)
+	common := fs.Filesystem()
+	i := 1
+	for {
+		_, err = os.Stat(filepath.Join(common.Root(), "worktrees", name, "gitdir"))
+		if os.IsNotExist(err) {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		name = filepath.Base(path) + strconv.Itoa(i)
+		i++
+	}
+
+	wt := osfs.New(path)
+	dot := osfs.New(filepath.Join(common.Root(), "worktrees", name))
+	repositoryFs := dotgit.NewRepositoryFilesystem(dot, common)
+	s := filesystem.NewStorage(repositoryFs, cache.NewObjectLRUDefault())
```

Comment:
I would add this as a new abstraction (e.g. `storer.WorktreeStorer`) within the [Storer](https://github.com/go-git/go-git/blob/master/storage/storer.go) interface. It would be up to the storage to handle the implementation details.